src/query/query_hash_scan.c

```diff
@@ -0,0 +1,421 @@
+  return DB_EQ;
+}
+
+/*
+ * qdata_hscan_key_eq () - check equality of two aggregate keys
+ *   returns: true if equal, false otherwise
+ *   key1(in): first key
+ *   key2(in): second key
+ */
+int
+qdata_hscan_key_eq (const void *key1, const void *key2)
+{
+  int decoy;
+
+  /* compare for equality */
+  return (qdata_hscan_key_compare ((HASH_SCAN_KEY *) key1, (HASH_SCAN_KEY *) key2, &decoy) == DB_EQ);
+}
+
+/*
+ * qdata_build_hscan_key () - build aggregate key structure from reguvar list
+ *   returns: NO_ERROR or error code
+ *   thread_p(in): thread
+ *   key(out): aggregate key
+ *   regu_list(in): reguvar list for fetching values
+ */
+int
+qdata_build_hscan_key (THREAD_ENTRY * thread_p, val_descr * vd, REGU_VARIABLE_LIST regu_list, HASH_SCAN_KEY * key)
+{
+  int rc = NO_ERROR;
+
+  /* build key */
+  key->free_values = false;	/* references precreated DB_VALUES */
+  key->val_count = 0;
+  while (regu_list != NULL)
+    {
+      rc = fetch_peek_dbval (thread_p, &regu_list->value, vd, NULL, NULL, NULL, &key->values[key->val_count]);
+      if (rc != NO_ERROR)
+	{
+	  return rc;
+	}
+
+      /* next */
+      regu_list = regu_list->next;
+      key->val_count++;
+    }
+
+  /* all ok */
+  return NO_ERROR;
+}
+
+/*
+ * qdata_print_hash_scan_entry () - Print the entry
+ *                              Will be used by mht_dump() function
+ *   return:
+ *   fp(in)     :
+ *   key(in)    :
+ *   data(in)   :
+ *   args(in)   :
+ */
+int
+qdata_print_hash_scan_entry (THREAD_ENTRY * thread_p, FILE * fp, const void *key, void *data, void *args)
+{
+  HASH_SCAN_VALUE *data2 = (HASH_SCAN_VALUE *) data;
+  HASH_SCAN_KEY *key2 = (HASH_SCAN_KEY *) key;
+  QFILE_TUPLE_VALUE_TYPE_LIST *list = (QFILE_TUPLE_VALUE_TYPE_LIST *) args;
+
+  if (!data2 || !key2)
+    {
+      return false;
+    }
+  if (fp == NULL)
+    {
+      fp = stdout;
+    }
+
+  fprintf (fp, "LIST_CACHE_ENTRY (%p) {\n", data);
+  fprintf (fp, "data_size = [%d]  data = [%.*s]\n", QFILE_GET_TUPLE_LENGTH (data2->tuple),
+	   QFILE_GET_TUPLE_LENGTH (data2->tuple), data2->tuple);
+
+  fprintf (fp, "key : ");
+  for (int i = 0; i < key2->val_count; i++)
+    {
+      db_fprint_value (fp, key2->values[i]);
+      fprintf (fp, " ");
+    }
+  fprintf (fp, "\n}");
+
+  return true;
+}
+
+/*
+ * qdata_copy_hscan_key () - deep copy aggregate key
+ *   returns: pointer to new aggregate hash key
+ *   thread_p(in): thread
+ *   key(in): source key
+ */
+HASH_SCAN_KEY *
+qdata_copy_hscan_key (cubthread::entry * thread_p, HASH_SCAN_KEY * key, REGU_VARIABLE_LIST probe_regu_list,
+		      val_descr * vd)
+{
+  HASH_SCAN_KEY *new_key = NULL;
+  int i = 0;
+  DB_TYPE vtype1, vtype2;
+  TP_DOMAIN_STATUS status = DOMAIN_COMPATIBLE;
+
+  if (key)
+    {
+      /* make a copy */
+      new_key = qdata_alloc_hscan_key (thread_p, key->val_count, false);
+    }
+
+  if (new_key)
+    {
+      /* copy values */
+      new_key->val_count = key->val_count;
+      for (i = 0; i < key->val_count; i++)
```

Comment:
I suggest a modified code to check null as below
`for (i = 0; i < key->val_count && probe_regu_list; i++)`

Reply:
I think there is no need to change it because it is checked in the check_hash_list_scan() function.
```
for (build_cnt = 0; build && probe; build_cnt++)
    {
      build = build->next;
      probe = probe->next;
    }
  /* The number of probe regu_var and build regu match */
  if (build != NULL || probe != NULL)
    {
      return false;
    }
  *val_cnt = build_cnt;
```